### PR TITLE
feat(web-fonts): add local font support to web font preset

### PR DESCRIPTION
**Pull request by onmax**
🏗️ WIP

Solves #3646

### CODE REVIEW

File: packages/preset-web-fonts/src/local-font.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+  async function saveFont(url: string, path: string) {
+    const response = await fetch!(url, { headers: { responseType: 'arraybuffer' } }) as ArrayBuffer
+    const content = new Uint8Array(response)
+    await writeFile(path, Buffer.from(content))
+    return { url, path }
+  }
+
+  // FIXME Currently hardcoded to the playground folder
+  const fontsFolder = resolve(dir, '../../../playground/assets/fonts')
```
Comment (onmax) [inline packages/preset-web-fonts/src/local-font.ts:39]: I need to write the new fonts into the assets folder but I am not sure how to get that **relative** path from the 'virtual:uno.css'

Any ideas?

File: playground/uno.config.ts
Comment (onmax) [inline playground/uno.config.ts:1]: When this PR is done I will revert this changes. Just for testing!

**Comment by onmax** [comment]
I have no idea how to get the path from the root project in order to write the fonts to the assets folder. 

I don't even know if this is the right solution...

Or should I use a virtual Vite file instead? I am blocked now 😓


Any ideas are much welcomed! Thanks!

**Comment by userquin** [comment]
I guess you will need to add an option to the preset  (where to download the fonts?, by default `${process.cwd()}/public`, sveltekit should be `${process.cwd()}/static`) and also a plugin to serve fonts: you also need to expose that option in the context, and so the plugin can resolve it. Check preset icons, we also have some option in the preset to resolve local collections.

I don't see any protection when running in the browser, you should protect the preset: check [preset-icons](https://github.com/unocss/unocss/blob/main/packages/preset-icons/src/core.ts#L177)

**Comment by userquin** [comment]
I'll try to add initial logic for vite (with small preset changes), later we can add webpack and CLI support (CLI will not require any integration since it is not a bundler, we only need to download logic): we need to add support for dev server to download the fonts

**Comment by userquin** [comment]
@onmax shouldn't the css also be saved locally?

**Comment by onmax** [comment]
> @onmax shouldn't the css also be saved locally?

Why? Don't we want to let UnoCSS handle the CSS for us?

Once you have the fonts downloaded in your machine, the preset will no longer make requests to the font provider but we will always update the CSS `url()` imports to link to the local font file

**Comment by onmax** [comment]
#3723 3723

**Pull request closed without merging**